sdk/go/pulumi/context.go

```diff
@@ -820,6 +815,18 @@ func (ctx *Context) registerResource(
 		options.Parent = nil
 	}
 
+	_, custom := resource.(CustomResource)
+	if !custom && (remote || len(options.URN) > 0) {
+		// Remote components are marked as such as well as any rehydrated component,
+		// for dependency tracking purposes. During dependency expansion, local
+		// components aren't included as dependencies, rather, their descendants are.
+		// But rehydrated components won't have any children from the SDK's perspective,
+		// so we mark them as remote components so that they themselves are kept as,
+		// dependencies, which is consistent with what is done for "dependency resources"
+		// in `newDependencyResource`.
+		resource.markRemoteComponent()
```

Comment:
As a follow-up, we should consider renaming `remoteComponent`, `markRemoteComponent`, and `isRemoteComponent` to something else to indicate that the resource should be kept as a dependency, since we're using it for more than just remote components (dependency resources and now rehydrated resources).

Reply:
I went ahead and renamed these and updated their comments.